django/core/paginator.py

```diff
@@ -178,6 +183,140 @@ def get_elided_page_range(self, number=1, *, on_each_side=3, on_ends=2):
             yield from range(number + 1, self.num_pages + 1)
 
 
+class AsyncPaginator(BasePaginator):
+    def __init__(
+        self,
+        object_list,
+        per_page,
+        orphans=0,
+        allow_empty_first_page=True,
+        error_messages=None,
+    ):
+        super().__init__(
+            object_list, per_page, orphans, allow_empty_first_page, error_messages
+        )
+        self._cache_acount = None
+        self._cache_anum_pages = None
+
+    async def __aiter__(self):
+        page_range = await self.apage_range()
+        for page_number in page_range:
+            yield await self.apage(page_number)
+
+    async def avalidate_number(self, number):
+        """See Paginator.validate_number()."""
+        try:
+            if isinstance(number, float) and not number.is_integer():
+                raise ValueError
+            number = int(number)
+        except (TypeError, ValueError):
+            raise PageNotAnInteger(self.error_messages["invalid_page"])
+        if number < 1:
+            raise EmptyPage(self.error_messages["min_page"])
+        num_pages = await self.anum_pages()
+        if number > num_pages:
+            raise EmptyPage(self.error_messages["no_results"])
+        return number
+
+    async def aget_page(self, number):
+        """See Paginator.get_page()."""
+        try:
+            number = await self.avalidate_number(number)
+        except PageNotAnInteger:
+            number = 1
+        except EmptyPage:
+            number = await self.anum_pages()
+        return await self.apage(number)
+
+    async def apage(self, number):
+        """See Paginator.page()."""
+        number = await self.avalidate_number(number)
+        bottom = (number - 1) * self.per_page
+        top = bottom + self.per_page
+        count = await self.acount()
+        if top + self.orphans >= count:
+            top = count
+        if not isinstance(self.object_list, list):
+            self.object_list = await sync_to_async(list)(self.object_list)
+            object_list = self.object_list[bottom:top]
+        else:
+            object_list = self.object_list[bottom:top]
+
+        return self._get_apage(object_list, number, self)
+
+    def _get_apage(self, *args, **kwargs):
+        """See Paginator.get_apage()."""
+        return AsyncPage(*args, **kwargs)
```

Comment:
I would move the `_get_page` hook to the `BasePaginator` class and overwrite this as just `_get_page`. When we add an a prefix it looks like something we need to await